2.10.1 Release Notes (#16340)
apache/pulsar: site2/website/release-notes.md

- [improve][broker] Add log when update namespace policies with error [14850](https://github.com/apache/pulsar/pull/14850)
- [feature][broker] Support advertised listeners for HTTP and HTTPS services [14839](https://github.com/apache/pulsar/pull/14839)
- [fix][broker] Filter the virtual NIC with relative path [14829](https://github.com/apache/pulsar/pull/14829)
- [fix][broker] Fixed duplicated delayed messages when all consumers disconnect [14740](https://github.com/apache/pulsar/pull/14740)
- [fix][broker] Fix cannot delete namespace with system topic [14730](https://github.com/apache/pulsar/pull/14730)
- [fix][broker] Fixed 404 error msg not being returned correctly using http lookup [14677](https://github.com/apache/pulsar/pull/14677)
- [fix][broker] Fix normal topic named ends with `healthcheck`  becomes system topic issue [14671](https://github.com/apache/pulsar/pull/14671)
- [improve][broker] Support shrink for map or set [14663](https://github.com/apache/pulsar/pull/14663)
- [improve][broker] Changing the topic creation flow and optimize heartbeat topic not trigger compaction [14643](https://github.com/apache/pulsar/pull/14643)

> - [fix][broker] Fix wrong prompt exception when get non-persistent topic list without GET_BUDNLE permission [14638](https://github.com/apache/pulsar/pull/14638)

Reviewer: ```suggestion
- [fix][broker] Fix wrong prompt exception when getting the non-persistent topic list without GET_BUDNLE permission [14638](https://github.com/apache/pulsar/pull/14638)
```

- [fix][txn] Fix NPE of TransactionMetaStoreHandler [15840](https://github.com/apache/pulsar/pull/15840)
- [improve][txn] Optimize transaction lowWaterMark to clean useless data faster [15592](https://github.com/apache/pulsar/pull/15592)
- [fix][txn] Fix transaction PendingAck lowWaterMark [15530](https://github.com/apache/pulsar/pull/15530)
- [fix][txn] Make transaction stats consistent at end txn [15472](https://github.com/apache/pulsar/pull/15472)
- [improve][txn] Add lowWaterMark check before appending entry to TB [15424](https://github.com/apache/pulsar/pull/15424)
- [fix][txn] Fix transaction component recover fillQueue [15418](https://github.com/apache/pulsar/pull/15418)
- [fix][txn] Topic transaction buffer recover don't close reader when throw RuntimeException [15361](https://github.com/apache/pulsar/pull/15361)
- [fix][txn] Fix potentially unfinishable future [15208](https://github.com/apache/pulsar/pull/15208)
- [improve][txn] Optimize metadataPositions in MLPendingAckStore [15137](https://github.com/apache/pulsar/pull/15137)
- [fix][txn] TransactionMetadataService don't connect again if store exist [15114](https://github.com/apache/pulsar/pull/15114)

> - [improve][txn] Avoid create multiple future and exception handler. [15089](https://github.com/apache/pulsar/pull/15089)

Reviewer: ```suggestion
- [improve][txn] Avoid creating multiple future and exception handlers [15089](https://github.com/apache/pulsar/pull/15089)
```

- [fix][txn] Fix transaction buffer no snapshot close recover reader [14830](https://github.com/apache/pulsar/pull/14830)
- [fix][txn] Fix transaction log recover throw cursor already close [14810](https://github.com/apache/pulsar/pull/14810)
- [fix][txn] Fix cannot enable transaction when is allow auto update schema enabled=fasle [14809](https://github.com/apache/pulsar/pull/14809)
- [fix][txn] Fix some exception handle in transaction buffer [14808](https://github.com/apache/pulsar/pull/14808)
- [fix][txn] Fix transaction buffer recover throw cursor already close [14807](https://github.com/apache/pulsar/pull/14807)
- [fix][txn] Fix transaction buffer recover reader and writer fail [14801](https://github.com/apache/pulsar/pull/14801)
- [fix][txn] Fix pending ack is recovering throw CursorAlreadyClosedxception [14781](https://github.com/apache/pulsar/pull/14781)
- [fix][txn] Fix cursor readPosition is bigger than maxPosition in OpReadEntry [14667](https://github.com/apache/pulsar/pull/14667)

### Clients

- [fix][Java] Fixes NPE when TableView handles null value message [15951](https://github.com/apache/pulsar/pull/15951)

> - [fix][Python] Fixes reserved keys is not removed when JsonSchema being encoded [15947](https://github.com/apache/pulsar/pull/15947)

Reviewer: ```suggestion
- [fix][Python] Fixes reserved keys are not removed when JsonSchema is being encoded [15947](https://github.com/apache/pulsar/pull/15947)
```

- [fix][Java] Fix conversion of `TimestampMillisConversion` has no effect when Jsr310Conversion enabled [15863](https://github.com/apache/pulsar/pull/15863)
- [fix][Python] Fix generated Python protobuf code not compatible with latest protobuf package [15846](https://github.com/apache/pulsar/pull/15846)
- [fix][Java] Fix messages sent by producers without schema cannot be decoded [15622](https://github.com/apache/pulsar/pull/15622)
- [fix][C++] avoid race condition causing double callback on close [15508](https://github.com/apache/pulsar/pull/15508)
- [improve][Java] improve logic when ACK grouping tracker checks duplicated message id [15465](https://github.com/apache/pulsar/pull/15465)

> - [improve][Java] Add pending messages information while print the producer stats [15440](https://github.com/apache/pulsar/pull/15440)

Reviewer: ```suggestion
- [improve][Java] Add pending messages information while printing the producer stats [15440](https://github.com/apache/pulsar/pull/15440)
```

- [fix][C++] Wait until event loop terminates when closing the Client [15316](https://github.com/apache/pulsar/pull/15316)
- [fix][Java] Fix negative ack not redelivery [15312](https://github.com/apache/pulsar/pull/15312)
- [improve][admin/client] AsyncHttpConnector doesn't use the system properties configured [15307](https://github.com/apache/pulsar/pull/15307)
- [fix][Java] TableView should cache created readers [15178](https://github.com/apache/pulsar/pull/15178)
- [fix][Java] Fix performance regression with message listener [15162](https://github.com/apache/pulsar/pull/15162)
- [fix][C++] Fix UnknownError might be returned for a partitioned producer [15161](https://github.com/apache/pulsar/pull/15161)
- [fix][C++] Fix single message metadata not set correctly [15072](https://github.com/apache/pulsar/pull/15072)

> - [improve][Python] Generate universal2 wheel files for MacOS [15054](https://github.com/apache/pulsar/pull/15054)

Reviewer: ```suggestion
- [improve][Python] Generate universal2 wheel files for macOS [15054](https://github.com/apache/pulsar/pull/15054)
```

> - [improve][Python] Added build for wheels on Alpine linux [15016](https://github.com/apache/pulsar/pull/15016)

Reviewer: ```suggestion
- [improve][Python] Added build for wheels on Alpine Linux [15016](https://github.com/apache/pulsar/pull/15016)
```

### Pulsar IO and Pulsar Functions

- [fix][connector] KCA sinks: fix offset mapping when sanitizeTopicName=true [15950](https://github.com/apache/pulsar/pull/15950)
- [improve][function] provide default error handler for function log appender [15728](https://github.com/apache/pulsar/pull/15728)
- [fix][connector] KCA Sink: org.apache.kafka.connect.errors.DataException: Invalid Java object for schema with type .. [15598](https://github.com/apache/pulsar/pull/15598)
- [improve][function] Refine file io connector [15250](https://github.com/apache/pulsar/pull/15250)
- [fix][function] Check executor null when close the FileSource [15247](https://github.com/apache/pulsar/pull/15247)
- [fix][function] Handle NPE when `getLeader` returns null [15058](https://github.com/apache/pulsar/pull/15058)
- [fix][function] Allow a Function<GenericObject,?> to access the original Schema of the Message and use it [14847](https://github.com/apache/pulsar/pull/14847)
- [fix][function] Fix pulsar-managed runtimes failed start function with package URL from package management service  [14814](https://github.com/apache/pulsar/pull/14814)

> - [improve][connector] Handle kafka sinks that return immutable maps as configs [14780](https://github.com/apache/pulsar/pull/14780)

Reviewer: ```suggestion
- [improve][connector] Handle Kafka sinks that return immutable maps as configs [14780](https://github.com/apache/pulsar/pull/14780)
```

- [improve][connector] Support event-time-based index name in ES Sink [14383](https://github.com/apache/pulsar/pull/14383)

### Tiered Storage

- [feature][tiered-storage] Add pure S3 provider for the offloader [15710](https://github.com/apache/pulsar/pull/15710
- [improve][tiered-storage] Upgrade JClouds to 2.5.0 [15649](https://github.com/apache/pulsar/pull/15649)

> - [improve][tiered-storage] Reduce cpu usage when offloading the ledger [15063](https://github.com/apache/pulsar/pull/15063)

Reviewer: ```suggestion
- [improve][tiered-storage] Reduce CPU usage when offloading the ledger [15063](https://github.com/apache/pulsar/pull/15063)
```

- [fix][authn] Fix typo checkPermissionsAsync method typo [15273](https://github.com/apache/pulsar/pull/15273)
- [improve][authn] Improve skipping of DNS resolution when creating AuthenticationDataHttp instance [15228](https://github.com/apache/pulsar/pull/15228)
- [improve][authn] Skip unnecessary DNS resolution when creating AuthenticationDataHttp instance [15221](https://github.com/apache/pulsar/pull/15221)
- [fix][security] Remove log4j for CVE-2022-23307 [15109](https://github.com/apache/pulsar/pull/15109)
- [improve][authn] Use tlsCertRefreshCheckDurationSec instead of 0 for refresh value [15075](https://github.com/apache/pulsar/pull/15075)
- [fix][security] Upgrade Spring Context in Pulsar IO batch-data-generator [14975](https://github.com/apache/pulsar/pull/14975)
- [fix][auth] Athenz: do not use uber-jar and bump to 1.10.50 to remove jackson-databind shaded dependency [14884](https://github.com/apache/pulsar/pull/14884)
- [fix][authz] Fix handling single role and non-jwt-token in MultiRolesTokenAuthorizationProvider [14857](https://github.com/apache/pulsar/pull/14857)

> - [improve][authn] Full-support ssl provider, ciphers and protocols for broker service and proxy service [14569](https://github.com/apache/pulsar/pull/14569)

Reviewer: ```suggestion
- [improve][authn] Full-support SSL provider, ciphers, and protocols for broker service and proxy service [14569](https://github.com/apache/pulsar/pull/14569)
```

- [fix][authz] Role with namespace produce authz can also get topics [13773](https://github.com/apache/pulsar/pull/13773)
- [improve][authn] Full-support set ssl provider, ciphers and protocols for broker-web&websocket/proxy/function-worker [13740](https://github.com/apache/pulsar/pull/13740)

### Others

> - [improve] Enable TCP/IP keepalive for all ZK client connections in all components started with bin/pulsar [15908](https://github.com/apache/pulsar/pull/15908)

Reviewer: ```suggestion
- [improve] Enable TCP/IP keepalive for all ZK client connections in all components starting with bin/pulsar [15908](https://github.com/apache/pulsar/pull/15908)
```

- [improve] Allow pulsar_tool_env.sh PULSAR_MEM to be Overridden [15868](https://github.com/apache/pulsar/pull/15868)

> - [improve] Configure DLog Bookie, Pulsar, and Admin clients via pass through config [15818](https://github.com/apache/pulsar/pull/15818)

Reviewer: ```suggestion
- [improve] Configure DLog Bookie, Pulsar, and Admin clients via pass-through config [15818](https://github.com/apache/pulsar/pull/15818)
```

- [fix][owasp] Fix false positive google-http-client-gson-1.41.0.jar [15651](https://github.com/apache/pulsar/pull/15651)
- [fix][webSocket] Fix MultiTopicReader#getConsumer ClassCastException [15534](https://github.com/apache/pulsar/pull/15534)
- [fix][package-management] Fix the new path `/data` introduced regression [15367](https://github.com/apache/pulsar/pull/15367)
- [improve][common] Use `Collection` instead of `List` for FutureUtil. [15329](https://github.com/apache/pulsar/pull/15329)
- [fix][tools] Only apply maxPendingMessagesAcrossPartitions if it presents [15283](https://github.com/apache/pulsar/pull/15283)
- [fix][owasp] Suppress MariaDB false positives [15243](https://github.com/apache/pulsar/pull/15243)
- [improve][webSocket]Add KeyStore support in WebSocket, Function Worker HTTPS Servers  [15084](https://github.com/apache/pulsar/pull/15084)
- [fix][scripts] Ignore case when obfuscating passwords in configuration scripts [15077](https://github.com/apache/pulsar/pull/15077)
- [improve][tool] Improve transaction perf logs [14816](https://github.com/apache/pulsar/pull/14816)

> - [fix][deploy] Fix the pid occupied check when use pulsar-daemon start or stop process [14701](https://github.com/apache/pulsar/pull/14701)

Reviewer: ```suggestion
- [fix][deploy] Fix the pid occupied check when using pulsar-daemon to start or stop process [14701](https://github.com/apache/pulsar/pull/14701)
```

- [fix][webSocket] Fix `ClassCastException` when user create `MultiTopicReader` [14316](https://github.com/apache/pulsar/pull/14316)

### Library updates

> - Bump pyyaml from 5.3.1 to 5.4.1 to solve CVE-2020-14343 [15989](https://github.com/apache/pulsar/pull/15989)

Reviewer: ```suggestion
- Bump PyYAML from 5.3.1 to 5.4.1 to solve CVE-2020-14343 [15989](https://github.com/apache/pulsar/pull/15989)
```
